### ISSUE TITLE
Not able to get desired performance from FastAPI

### ISSUE THREAD

**Issue by jayudhandha** [tiangolo/fastapi#5316]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Union
from fastapi import FastAPI
import uvicorn

app = FastAPI()

@app.get("/")
def read_root():
    return {"Hello": "World"}

@app.get("/items/{item_id}")
def read_item(item_id: int, q: Union[str, None] = None):
    return {"item_id": item_id, "q": q}

if __name__ == '__main__':
    uvicorn.run(app, port=8080, host='0.0.0.0')
```

### Description

* Deploy above code on GCP Cloud run
* Invoke GET API from apache JMeter
* 1000 request/sec for 180 sec duration

### Operating System

Windows

### Operating System Details

Intel Core i7, 16GB RAM

### FastAPI Version

0.68.0

### Python Version

3.8

### Additional Context

I tried above simple API in **Flask** & **FastAPI** and observed that performance was almost similar. I expect better performance in FastAPI as mentioned in their documentation.

My throughput in **apache jMeter** for **Flask** & **fastAPI** was similar like **200 req/sec** if run on 1 instance in cloud run.

Anyone did similar kind of exercise or comparision?

**Comment by iudeen** [comment]
Few suggestions:

1. Upgrade your FastAPI versions
2. Make sure you install uvicorn[standard]

Few questions:

How have you hosted? Do you have nginx? Is it docker?

**Comment by jayudhandha** [comment]
Hi @iudeen thanks for suggestions.

I Tried, But no improvements.

**Comment by zhanymkanov** [comment]
One of the ways to optimize the performance is to replace `def` routes with `async def` since former ones are called in a thread pool, while async routes are called directly.

Check it in [docs](https://fastapi.tiangolo.com/async/#path-operation-functions)

> When you declare a path operation function with normal def instead of async def, it is run in an external threadpool that is then awaited, instead of being called directly (as it would block the server).

> If you are coming from another async framework that does not work in the way described above and you are used to define trivial compute-only path operation functions with plain def for a tiny performance gain (about 100 nanoseconds), please note that in FastAPI the effect would be quite opposite. In these cases, it's better to use async def unless your path operation functions use code that performs blocking I/O.

**Comment by jayudhandha** [comment]
Yes i tried with `async` keyword as well, but nothing much difference in performance.

May be i will try to dig more into it once we start actual development.

Thanks guys for your inputs!

**Comment by yinziyan1206** [comment]
cancel uvicorn.access logging make 2x performance

**Comment by Kludex** [comment]
How are you benchmarking it?

**Comment by jayudhandha** [comment]
@Kludex I am using apache Jmeter

**Comment by jayudhandha** [comment]
@yinziyan1206 will give a try for sure.

**Comment by KalebRoncatti-ciandt** [comment]
As you are deploying to a cloud environment, you could also mess around with **_workers_** as explained on the docs:
- [FastAPI docs](https://fastapi.tiangolo.com/deployment/server-workers/)
- [Uvicorn docs](https://www.uvicorn.org/deployment/)

When running directly from Python you are probably not taking advantage of parallelization.

**Comment by DavidBord** [comment]
@jayudhandha I tried to reproduce your benchmark but with ab instead of jMeter.
ab accepts -c argument for concurrency.
In my setup, c equals 1 gets me to ~250 requests per second.
c equals 10 gets me to ~490 requests per second.
Please try to find a similar concurrency configuration for jMeter / switch to ab.

Switching `def` to `async def` leaves us with the same numbers.

Now, as @KalebRoncatti-ciandt running the app with --workers 10 brings to to ~1200 requests per second.

**Comment by raphaelauv** [comment]
replace 

```python
    return {"item_id": item_id, "q": q}
```

by

```python
    return ORJSONResponse(content={"item_id": item_id, "q": q})
```

it will disable the check on the response and use ORJSON for the serialization
